### DISCUSSION TITLE
Join forces with zxing-cpp?

### DISCUSSION THREAD

**Discussion by axxel** [micjahn/ZXing.Net#439] [General]
Hi, I'm the maintainer of https://github.com/zxing-cpp/zxing-cpp, the most recent/active c++ port of the Java code. I was browsing through the heap of individual ports (among them: Qt, ObjectiveC, JS, .NET) and noticed that all of them ported the Java code into their native language thereby basically duplicating the code with all its bugs and limitations. I'd love to help consolidate the rather fragmented landscape of zxing ports and bring the community closer together working on one solution instead of fixing issues multiple times in different code bases.

I spent a considerable amount of time over the last 4 years improving/reimplementing the underlying algorithms. The code is now 2x to 10x faster and also in part substantially more capable and reliable (detect more codes with less false positives). There are discussions to merge efforts with the zint project to substantially improve the writing/generating capabilities as well. 

I noticed that this project here is one of the few that are still actively maintained but has to deal with a number of issues that are (most likely) solved in zxing-cpp. Here is a random list I came up with after skimming though the issues: #390, #322, #326, #375, #345, #384.

So I'd like to propose to join forces by bringing your .NET expertise to zxing-cpp in form of a suitable wrapper (joining the set we already have: Android, iOS, Python, WASM, WinRT). The interface would be much simplified (see the _one_ function [ReadBarcodes()](https://github.com/zxing-cpp/zxing-cpp/blob/master/core/src/ReadBarcode.h)) but it would result in an API breaking change to your client code, which is obviously something to consider.

I'd love to hear your opinion. In case you find this proposal offensive, I apologize in advance.

**Comment by micjahn** [comment]
Thanks for your offer. I understand your point of view.
But ZXing.Net is one of my babies. ;) I started this project in 2012.
I want to keep it alive as long as there are users who include it in their projects. I hope you understand what I mean.

And perhaps it could be a better way to have both, the native ZXing.Net and a simple wrapper implementation for ReadBarcodes calls to zxing-cpp. Other developers are free to choose between both implementations. Btw. .Net developers should already be able to use your code by P/Invoke calls to ReadBarcodes.

**Reply by axxel** [reply]
Yes, I do understand. :).

I assume that means if I wanted to provide an out-of-the-box .NET experience (probably based on P/Invoke), I'd need to find someone else to provide that? ;)

**Reply by micjahn** [reply]
Maybe not. I would also work on two projects with a similar goal.
But please don't expect a solution within the next three days.
I have to do this in my spare time.

**Reply by axxel** [reply]
Of course. Thanks for considering it. Looking forward to your input on the matter.

**Reply by axxel** [reply]
I just stumbled upon this discussion again. Since you did not follow up on the idea yet, I assume you have either no time or no interest in the proposed collaboration?

**Reply by micjahn** [reply]
It is more or less a time and priority thing. I started an implementation last year which doesn't instantly work. And other important things moved in. I'm not sure when I will find enough free time and motivation to start a new attempt. Sorry.

**Reply by axxel** [reply]
I completely understand. Thanks for sharing your thoughts.

**Comment by axxel** [comment]
I decided to implement the [.NET wrapper](https://github.com/zxing-cpp/zxing-cpp/tree/master/wrappers/dotnet) for zxing-cpp myself. And since this made it very easy to throw together a little benchmark comparison tool (and I wanted to proof to myself that the couple days I put into this were worth it ;)), here are the [results](https://github.com/axxel/zxing-bench/tree/main/dotnet).

As a bit of a constructive critique, I have to say that I found your project lacking in the "quick start" department of your documentation. There seems to be ample inline documentation in your source code but it took me quite a while until I had [this function](https://github.com/axxel/zxing-bench/blob/2c52c15aa86fc95c34ad52e8798c00d578c62416/dotnet/Program.cs#L25) gobbled together from different parts of your source tree.

**Reply by axxel** [reply]
Quick update: I changed the benchmarking tool to use SkiaSharp instead of ImageMagic to fix some crashes in your (and my) code caused by png images with some weird 4-bit pixel formats. I also added the ability to let the tool run over a whole directory of images. I updated the readme accordingly.

I found it indeed surprising that my .NET wrapper around zxing-cpp turned out to be between **10x and 50x faster** compared to your solution. I'm wondering whether I did something "stupid" in how I use your code. In case you find something fishy with [this function](https://github.com/axxel/zxing-bench/blob/4bc298343076637d47838684dde46aa28d8933d1/dotnet/Program.cs#L50), please let me know.

**Reply by micjahn** [reply]
Great results. Thanks for letting me know.
I'm not surprised by the speed differences. As far as I know, you spent a lot of time in rewriting and optimizing the original Java-based zxing code base in C++. I think your implementation doesn't need the additional options like "TRY_HARDER", "AUTO_ROTATE" or "ALSO_INVERTED" to get better results than the original implementation. The fastest implementation would be the following
```
        var reader = new ZXing.SkiaSharp.BarcodeReader();
        return (filename) =>
        {
            using (var img = SKBitmap.Decode(filename))
            {
                var source = new SKBitmapLuminanceSource(img);
                //return reader.DecodeMultiple(source)?.Length ?? 0;
                return reader.Decode(source) == null ? 0 : 1;
            }
        };
```
But it doesn't look for multiple codes in one image and finds only 871 codes in 1253 files. Your cpp implementation finds 1004 codes.

In my opinion, the speed isn't the problem. There are too many misses.

**Reply by axxel** [reply]
> Great results. Thanks for letting me know. I'm not surprised by the speed differences. As far as I know, you spent a lot of time in rewriting and optimizing the original Java-based zxing code base in C++.

That is correct.

> I think your implementation doesn't need the additional options like "TRY_HARDER", "AUTO_ROTATE" or "ALSO_INVERTED" to get better results than the original implementation.

That is also true but I don't understand what you are trying to say here. Those options are still 'on' by default in my code base because they increase the number of successful detections.

> The fastest implementation would be the following [...]

I added a `--single` mode to my benchmarking tool. This made both our versions about twice as fast compared to without (see [here](https://github.com/axxel/zxing-bench/blob/main/dotnet/README.md)).

> But it doesn't look for multiple codes in one image and finds only 871 codes in 1253 files. Your cpp implementation finds 1004 codes.

Which "files" are you talking about here?

> In my opinion, the speed isn't the problem. There are too many misses.

You mean detection rate is more important than runtime (in general)?

**Comment by Laurnz** [comment]
Hey, I have a question on what the options would be for cooperation, as I am just a CS student and interested on how such things work in practice. Thank you both for creating and maintaining these awesome libraries.

I have tested both your implementations and must say that I have very hard images for 2D code detection.
I would love to use an implementation that is just pure C#, as I don't have to deal with native dependencies. But the C++ version detects the codes much better on different rotations and positions in the image, that I will use it going forward.

But since the pure C# version is also a project that @micjahn most likely does not want to stop, means that joining forces will not mean working on the same repository. So do you just look how the other one implemented some of the detectors to also get some ideas for improvement on the own repository? I would imagine that it would be very time-consuming to get into the details of the other repo.

Would you even want to implement the same principles for the detectors, just in different languages, or is the fun coming up with your own algorithms?
Or would you share design decisions of the algorithms in an understandable non code format for the others to better reason about and improve?

These might be some stupid questions, but I myself would be interested in knowing why some design decisions in the algorithms have been made and what improvements they lead to, as they are very complicated and have a lot of experience encoded in them.

**Reply by micjahn** [reply]
ZXing was originally written in java. ZXing.Net is a port to .Net which I made several years ago. I didn't have the goal to understand all of the java code nor to rewrite all of that stuff from scratch. I throw away all the exception-based flow control because it slows down the speed of an application a lot in .Net. Exceptions are only used if there occurs a real exceptional error. That is the only main difference between the Java and the .Net version.
I wish that Zxing.Net would be as fast and reliable as the cpp version. But to be realistic I don't have the free time to analyze and rewrite the code base in such a way.
On the other hand, I don't want to abandon my little baby which I have maintained for over 10 years now.
If there are others, who prefer pure C# solutions feel free to provide some PRs with improvements.
Anyway, is absolutely ok for me if the users of ZXing.Net are switching to zxing-cpp with the .Net wrapper to get a solution with better results.
